docs/proposals/interceptors.md

```diff
@@ -0,0 +1,196 @@
+# Proposal: Introduce GraphQL Interceptors
+
+_Owners_: @ThisaruGuruge @DimuthuMadushan     
+_Reviewers_: @shafreenAnfar @ThisaruGuruge       
+_Created_: 2022/06/07   
+_Updated_: 2021/07/18     
+_Issue_: [#2001](https://github.com/ballerina-platform/ballerina-standard-library/issues/2001)
+
+
+## Summary
+Introduce the GraphQL Interceptors to enhance the execution flow of the GraphQL services. The GraphQL interceptors can execute custom code before and after the resolver function gets invoked which helps with common stuff like authentication, authorization, logging, observability, etc.
+
+## Goals
+* Introduce service level interceptors
+* Introduce resolver level interceptors
+
+## Non-Goals
+* Add Interceptor support for each event in the response stream of subscription
+
+## Motivation
+
+Even though the Ballerina GraphQL package provides almost all the main features in GrpahQL spec, there are a few more additional features that need to be added to enhance the execution flow and usability. Due to the unavailability of these additional features, it’s difficult to perform some useful operations (Ex: Observability) with the existing capabilities of the GraphQL package. When it comes to popular GraphQL packages, they have introduced the [Middleware](https://github.com/maticzav/graphql-middleware#readme) concept which addresses the above difficulties. However, the Middleware concept is very much similar to the Interceptor concept in the Ballerina HTTP package. Adapting the Interceptor concept into the GraphQL package will resolve the aforementioned limitations. With the GraphQL Interceptors, It will be possible to implement the following functionalities easily.
+* Authentication
+* Authorization
+* Observability
+* Logging
+* Input Sanitization
+* Output Manipulation
+
+In some GraphQL services, resolvers are getting cluttered with business logic which reduces code readability drastically. As an example, consider the Ballerina GraphQL services which use security stuff. The developer has to write the security logic inside the resolver along with the business logic. But the GraphQL interceptor provides a clear separation of concerns that improves the code modularity. At the same time, It eliminates the repetitive code from the resolvers and keeps the resolver clean and simple.
+
+## Description
+
+The Interceptor concept is much similar to the GraphQL middleware concept. The key attributes of interceptors are as follows.
+* Can execute any code before/after invoking a Resolver function
+* Can return errors
+* Have access to the `graphql:Context` and the `graphql:RequestInfo`
+* Can apply to multiple resolvers
+* Can pass state from one interceptor to another
+
+>NOTE: In the first phase, interceptors are supported only for `Query` and `Mutation` operations.
```

Comment:
We don't need this. The proposal should say what it is going to implement. Future plans can be a separate section if we want to mention them.